src/DotVVM.Framework.Tests.Common/ViewModel/ViewModelTypeMetadataSerializerTests.cs

```diff
@@ -0,0 +1,172 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Text;
+using DotVVM.Framework.Configuration;
+using DotVVM.Framework.ViewModel;
+using DotVVM.Framework.ViewModel.Serialization;
+using DotVVM.Framework.ViewModel.Validation;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace DotVVM.Framework.Tests.Common.ViewModel
+{
+    [TestClass]
+    public class ViewModelTypeMetadataSerializerTests
+    {
+        private static ViewModelSerializationMapper mapper;
+
+        [ClassInitialize]
+        public static void ClassInit(TestContext context)
+        {
+            mapper = new ViewModelSerializationMapper(new ViewModelValidationRuleTranslator(),
+                new AttributeViewModelValidationMetadataProvider(),
+                new DefaultPropertySerialization(),
+                DotvvmConfiguration.CreateDefault());
+        }
+
+        [DataTestMethod]
+        [DataRow(typeof(bool), "'Boolean'")]
+        [DataRow(typeof(int?), "{'type':'nullable','inner':'Int32'}")]
+        [DataRow(typeof(SampleEnum), "{'type':'enum','values':{'Zero':0,'One':1,'Two':2}}")]
+        [DataRow(typeof(long[][]), "[['Int64']]")]
+        [DataRow(typeof(Type), "'t38YPaufyA26odb7HXM9a+aghdA='")]   // unknown types should produce SHA1 hash
+        public void ViewModelTypeMetadata_TypeName(Type type, string expected)
+        {
+            var typeMetadataSerializer = new ViewModelTypeMetadataSerializer();
+            var result = typeMetadataSerializer.GetTypeIdentifier(type);
+            Assert.AreEqual(expected.Replace("'", "\""), result.ToString(Formatting.None));
+        }
+
+        [TestMethod]        
+        public void ViewModelTypeMetadata_TypeMetadata()
+        {
+            var typeMetadataSerializer = new ViewModelTypeMetadataSerializer();
+            var result = typeMetadataSerializer.SerializeTypeMetadata(new[]
+            {
+                mapper.GetMap(typeof(TestViewModel)),
+                mapper.GetMap(typeof(NestedTestViewModel))
+            });
+
+            var expected = @"{
+    'hXgHohhzHL2SzMTI0/aQFP42rV8=': {
+        'ChildFirstRequest': {
+            'type': 'GqKC8CuCoDNJdqzKjKrxePbUZr8=',
+            'post': 'no',
+            'update': 'firstRequest'
+        },
+        'ClientToServer': {
+            'type': 'String',
+            'update': 'no'
+        },
+        'NestedList': {
+            'type': [
+                'GqKC8CuCoDNJdqzKjKrxePbUZr8='
+            ]
+        },
+        'property ONE': {
+            'type': 'Guid'
+        },
+        'property TWO': {
+            'type': [
+                {
+                    'type': 'nullable',
+                    'inner': {
+                        'type': 'enum',
```

Comment:
Hmm, could we pre-declare the enums like we do for types? This could get quite long if you would use the enum often.